xbmc/platform/darwin/speech/SpeechRecognitionDarwin.mm

```diff
@@ -0,0 +1,331 @@
+                           case 1107: // Connection to speech process was interrupted
+                             break;
+                         }
+                       }
+
+                       NSString* logMsg =
+                           [NSString stringWithFormat:@"code='%ld' description='%@'",
+                                                      (long)error.code, error.localizedDescription];
+                       [sself onError:recognitionError logMessage:logMsg];
+                     }
+                     else
+                     {
+                       owner->OnRecognitionDone(listener);
+                     }
+
+                     sself.recognitionRequest = nil;
+                     sself.recognitionTask = nil;
+                     sself.text = nil;
+                     [sself.talkTimeoutTimer invalidate];
+                     sself.talkTimeoutTimer = nil;
+                   }];
+
+  [inputNode installTapOnBus:0
+                  bufferSize:4096
+                      format:[inputNode outputFormatForBus:0]
+                       block:^(AVAudioPCMBuffer* buffer, AVAudioTime* when) {
+                         [self.recognitionRequest appendAudioPCMBuffer:buffer];
+                       }];
+
+  [self.audioEngine prepare];
+
+  NSError* outError;
+  [self.audioEngine startAndReturnError:&outError];
+
+  if (outError != nil)
+  {
+    [self onError:speech::RecognitionError::AUDIO
+        logMessage:[NSString stringWithFormat:
+                                 @"Audio engine couldn't start because of an error. code='%ld'",
+                                 outError.code]];
+  }
+}
+
+- (void)speechRecognizer:(SFSpeechRecognizer*)speechRecognizer availabilityDidChange:(BOOL)available
+{
+  if (available)
+    return;
+
+  [self.recognitionTask cancel];
+  self.recognitionTask = nil;
+
+  if (self.listener != nil && self.owner != nil)
+  {
+    [self onError:speech::RecognitionError::SERVICE_NOT_AVAILABLE
+        logMessage:@"Service currently not available. Try again later"];
+  }
+}
+
+- (void)onTalkTimeout:(NSTimer*)timer
+{
+  [self.recognitionRequest endAudio];
+}
+
+- (void)onError:(int)recognitionError logMessage:(NSString*)logMessage
+{
+  CLog::Log(LOGERROR, "Speech recognition error: {}", [logMessage UTF8String]);
+  self.listener->OnError(recognitionError);
+  self.owner->OnRecognitionDone(self.listener);
+}
+
+@end
+
+std::shared_ptr<speech::ISpeechRecognition> speech::ISpeechRecognition::CreateInstance()
+{
+  return std::make_shared<CSpeechRecognitionDarwin>();
+}
+
+struct API_AVAILABLE(macos(10.15), ios(10.0)) API_UNAVAILABLE(tvos) SpeechRecognitionDarwinImpl
+{
+  CCriticalSection m_listenersMutex;
+  std::vector<std::shared_ptr<speech::ISpeechRecognitionListener>> m_listeners;
+  SpeechRecognitionImpl* m_recognizer{nil};
+};
+
+CSpeechRecognitionDarwin::CSpeechRecognitionDarwin() : m_impl(new SpeechRecognitionDarwinImpl)
```

Comment:
iirc it's preferred to use `std::make_unique<>`

Reply:
In fact, this is just a matter of taste and main argument is shorter code. But here it will make the code longer. I'll keep it as is.